### PR TITLE
Create replay lambda

### PR DESCRIPTION
**Pull request by connorIW**
This lambda will receive an original request and original response from the V2 get_award_details lambda and query the V1 equivalent Lambda to confirm that both return the same values.

### CODE REVIEW

File: src/replayer.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+
+def get_parameters():
+    parser = argparse.ArgumentParser(
+        description="Convert S3 objects into Kafka messages"
+    )
+
+    # Parse command line inputs and set defaults
+    parser.add_argument("--aws-profile", default="default")
+    parser.add_argument("--aws-region", default="eu-west-2")
```
Comment (dan-hill2802) [inline src/replayer.py:50]:
```suggestion
    parser.add_argument("--aws-region", default="eu-west-1")
```

File: src/replayer.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    if the_logger.isEnabledFor(logging.DEBUG):
+        boto3.set_stream_logger()
+        the_logger.debug(f'Using boto3", "version": "{boto3.__version__}')
+
+    return the_logger
+
+
+def get_parameters():
+    parser = argparse.ArgumentParser(
+        description="Convert S3 objects into Kafka messages"
```
Comment (dan-hill2802) [inline src/replayer.py:45]: I suspect this is a copy/paste error

File: src/replayer.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    )
+
+    # Parse command line inputs and set defaults
+    parser.add_argument("--aws-profile", default="default")
+    parser.add_argument("--aws-region", default="eu-west-2")
+    parser.add_argument("--environment", default="NOT_SET")
+    parser.add_argument("--application", default="NOT_SET")
+    parser.add_argument("--log-level", default="INFO")
+    parser.add_argument("--aws-region", default="eu-west-1")
```
Comment (dan-hill2802) [inline src/replayer.py:54]: you have a duplicate

File: src/replayer.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+    parser.add_argument("--api-hostname", default="NOT_SET")
```
Comment (dan-hill2802) [inline src/replayer.py:55]: this would be better without a default

File: src/replayer.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+                ", ".join(missing_args)
+            ),
+        )
+
+    return _args
+
+
+args = get_parameters()
+logger = setup_logging(args.log_level)
+region = os.environ.get("AWS_REGION")
```
Comment (dan-hill2802) [inline src/replayer.py]: this isn't needed as it is already in arg parser

File: src/replayer.py [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        f'Received response from AWS API", "api_hostname": "{args.hostname}", "response_code": "{request.status_code}'
+    )
+
+    return json.load(StringIO(request.text))
+
+
+def decrypt_response(response: dict, request: dict) -> dict:
+    # Create a deep copy of the response to keep the function pure
+    response = response.copy()
+    session = boto3.session.Session(profile_name="decrypt", region_name=region)
```
Comment (dan-hill2802) [inline src/replayer.py:168]:
```suggestion
    session = boto3.session.Session(profile_name="decrypt", region_name=args.aws_region)
```

File: src/replayer.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+
+    session = boto3.session.Session()
+    default_credentials = session.get_credentials().get_frozen_credentials()
+
+    loaded_event = json.loads(event)
+    try:
+        original_request = loaded_event["Body"]["originalRequest"]
+        original_response = loaded_event["Body"]["originalResponse"]
+    except Exception as e:
+        logger.error("Attempted to extract event items but was unable.")
```
Comment (dan-hill2802) [inline src/replayer.py:122]: currently you are continuing on error
```suggestion
    except KeyError as e:
        logger.error("Attempted to extract event items but was unable.")
        raise e
```

File: src/replayer.py [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+    nino = original_request.get("nino")
+    transaction_id = original_request.get("transaction_id")
+    from_date = original_request.get("from_date")
+    to_date = original_request.get("to_date")
```
Comment (dan-hill2802) [inline src/replayer.py]: these two might not exist according to our API docs

File: src/replayer.py [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    client = session.client("kms")
+
+    for period in response.get("assessmentPeriod", []):
+        amount = period.get("amount")
+
+        key_id = amount.get("keyId")
+        take_home_pay = base64.urlsafe_b64decode(amount.get("takeHomePay"))
+        cipher_text_blob = base64.urlsafe_b64decode(amount.get("cipherTextBlob"))
+
+        kms_response = client.decrypt(CiphertextBlob=cipher_text_blob, KeyId=key_id)
```
Comment (dan-hill2802) [inline src/replayer.py:179]: don't need key_id for KMS
```suggestion
        kms_response = client.decrypt(CiphertextBlob=cipher_text_blob)
```

File: src/replayer.py [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+            period["amount"] = amount
+
+        except Exception as e:
+            logger.error(
+                f'Failed to decrypt data", '
+                f'"transaction_id": {request.get("transaction_id")}, '
+                f'"from_date": {request.get("from_date")}, '
+                f'"to_date": {request.get("to_date")}'
+            )
+            logger.error(e)
```
Comment (dan-hill2802) [inline src/replayer.py:213]: I'm not sure you can continue if you fail to decrypt any of the take home pay, so should be treated as fatal
```suggestion
            raise e
```

File: src/replayer.py [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+
+    # Will return a copy of the response if claimantFound is False
+    # Will return a DECRYPTED copy of the response if claimantFound is True
```
Comment (dan-hill2802) [inline src/replayer.py:216]: I think these are no longer relevant

File: src/replayer.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+    except Exception as e:
+        logger.error("Attempted to extract event items but was unable.")
+
+    nino = original_request.get("nino")
+    transaction_id = original_request.get("transaction_id")
+    from_date = original_request.get("from_date")
+    to_date = original_request.get("to_date")
+
+    actual_response = replay_original_request(
+        default_credentials, nino, transaction_id, from_date, to_date
```
Comment (dan-hill2802) [inline src/replayer.py]: need to handle from_data and to_date not existing

File: src/replayer.py [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+    auth = AWSRequestsAuth(
+        aws_access_key=default_credentials.access_key,
+        aws_secret_access_key=default_credentials.secret_key,
+        aws_token=default_credentials.token,
+        aws_host=f"{args.api_hostname}",
+        aws_region=f"{args.aws_region}",
+        aws_service="execute-api",
+    )
+
+    request_parameters = f"nino={nino}&transactionId={transaction_id}&fromDate={from_date}&toDo={to_date}"
```
Comment (dan-hill2802) [inline src/replayer.py]: typo
```suggestion
    request_parameters = f"nino={nino}&transactionId={transaction_id}&fromDate={from_date}&toDate={to_date}"
```
also needs to be able to query without dates

File: src/replayer.py [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+
+    # Will return a copy of the response if claimantFound is False
+    # Will return a DECRYPTED copy of the response if claimantFound is True
+    return response
+
+
+def compare_responses(original, actual, request):
+    match = True
+    if not (original["claimantFound"] is True and actual["claimantFound"] is True):
+        match = False
```
Comment (dan-hill2802) [inline src/replayer.py:223]: if they are both false, it should be a match

File: src/replayer.py [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+        logger.error(
```
Comment (dan-hill2802) [inline src/replayer.py:224]: this should be at info level

File: src/replayer.py [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+            logger.error(e)
+
+    # Will return a copy of the response if claimantFound is False
+    # Will return a DECRYPTED copy of the response if claimantFound is True
+    return response
+
+
+def compare_responses(original, actual, request):
+    match = True
+    if not (original["claimantFound"] is True and actual["claimantFound"] is True):
```
Comment (dan-hill2802) [inline src/replayer.py:222]: a `==` comaparitor might be better

File: src/replayer.py [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+        match = False
+        logger.error(
+            f"Claimant found doesn't match, "
+            f'expected {original["claimantFound"]} from replayed response but got {actual["claimantFound"]}'
+        )
+
+    if original.get("suspendedDate"):
+        state = original["suspendedDate"] == actual["suspendedDate"]
+        if state is True:
```
Comment (dan-hill2802) [inline src/replayer.py:231]: what does this provide over 
```suggestion
        if original["suspendedDate"] == actual["suspendedDate"] is True:
```

File: src/replayer.py [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+            logger.info('Suspended date is a match", "status": "match')
```
Comment (dan-hill2802) [inline src/replayer.py:232]: are these `"` and `'` correct??

File: src/replayer.py [hunk 19]
```diff
@@ -0,0 +1,306 @@
...
+        else:
+            match = False
+            logger.info(
+                'Suspended date expected but not found in replayed response", "status": "miss'
```
Comment (dan-hill2802) [inline src/replayer.py:236]:
```suggestion
                'Suspended date expected but does not match or was not found in replayed response", "status": "miss'
```

**Review by steveburton4 — APPROVED** [review]

**Review by steveburton4 — APPROVED** [review]

**Pull request merged**